File: software diagrams/SoftwareIdeasModeler-15-51/Data/Templates/Code/Python/Default/Interface.py
<%==SIM:ExplicitWhitespacesOn==%>
<%==SIM:AutoIndentOn==%>
<%==SIM:UserInput:IndentationSize=Number("Indentation size",4)==%>
<%==SIM:UserInput:IndentationChar=Choice("Indentation char",Space,Tab)==%>
<%==SIM:If:Eq(IndentationChar,"Tab")==%><%==SIM:SetTabIndent==%><%==SIM:Else==%><%==SIM:SetSpaceIndent==%><%==SIM:EndIf==%><%==SIM:IndentSize(IndentationSize)==%>
<%==SIM:Set:private=__==%><%==SIM:Set:protected===%><%==SIM:Set:package===%><%==SIM:Set:public===%>

<%==SIM:ForEach:Element.DocumentationLines==%>#<%==SIM:DocumentationLine==%><%==SIM:Line==%>
<%==SIM:EndFor==%>
class <%==SIM:Element.Name==%><%==SIM:If:Element.HasSuperClassOrInterface==%>(<%==SIM:ForEach:Element.SuperClasses==%><%==SIM:SuperClass.Name==%><%==SIM:IfNot:IsLastItem==%>, <%==SIM:EndIf==%><%==SIM:EndFor==%><%==SIM:If:Element.HasSuperClass==%>,<%==SIM:EndIf==%> <%==SIM:ForEach:Element.Interfaces==%><%==SIM:Interface.Name==%><%==SIM:IfNot:IsLastItem==%>, <%==SIM:EndIf==%><%==SIM:EndFor==%>)<%==SIM:EndIf==%>:
<%==SIM:StartBlock==%><%==SIM:Line==%>
<%==SIM:ForEach:Element.AllAttributes==%><%==SIM:If:Attribute.IsStatic==%>    
	
	<%==SIM:ForEach:Attribute.DocumentationLines==%>
	#<%==SIM:DocumentationLine==%><%==SIM:Line==%>
    <%==SIM:EndFor==%>
	
	<%==SIM:Attribute.Visibility==%><%==SIM:Attribute.Name==%> = <%==SIM:If:Attribute.HasDefaultValue==%><%==SIM:Attribute.DefaultValue==%><%==SIM:EndIf==%><%==SIM:IfNot:Attribute.HasDefaultValue==%><%==SIM:If:Attribute.OriginalType.HasClassifier==%><%==SIM:Attribute.OriginalType.Classifier.Name==%>()<%==SIM:ElseIf:Attribute.OriginalType.IsString==%>""<%==SIM:ElseIf:Attribute.OriginalType.IsNumber==%>0<%==SIM:ElseIf:Attribute.OriginalType.IsBoolean==%>False<%==SIM:Else==%>None<%==SIM:EndIf==%><%==SIM:EndIf==%>
	<%==SIM:Line==%>

    assert(type(<%==SIM:Attribute.Visibility==%><%==SIM:Attribute.Name==%>)==<%==SIM:Attribute.Type==%>)
	<%==SIM:Line(2)==%>
<%==SIM:EndIf==%><%==SIM:EndFor==%>

    <%==SIM:Set:Constructor=Element.Constructors.First==%>
	def __init__(self<%==SIM:ForEach:Constructor.Parameters==%>, <%==SIM:Parameter.Name==%><%==SIM:EndFor==%>):	
	<%==SIM:StartBlock==%><%==SIM:Line==%>

	<%==SIM:ForEach:Element.AllAttributes==%><%==SIM:IfNot:Attribute.IsStatic==%>
            <%==SIM:IfNot:Attribute.HasStereotype("property")==%>
			    <%==SIM:ForEach:Attribute.DocumentationLines==%>
			    #<%==SIM:DocumentationLine==%><%==SIM:Line==%>	
			    <%==SIM:EndFor==%>
            <%==SIM:EndIf==%>

            <%==SIM:Set:Assignment===%>
            <%==SIM:ForEach:Constructor.Parameters==%>
            <%==SIM:If:Eq(Parameter.Name,Attribute.Name)==%>
            <%==SIM:Set:Assignment=Parameter.Name==%>
            <%==SIM:EndIf==%>
            <%==SIM:EndFor==%>


			self.<%==SIM:Attribute.Visibility==%><%==SIM:Attribute.Name==%> = <%==SIM:IfNot:IsEmpty(Assignment)==%><%==SIM:Assignment==%><%==SIM:Else==%><%==SIM:If:Attribute.HasDefaultValue==%><%==SIM:Attribute.DefaultValue==%><%==SIM:EndIf==%><%==SIM:IfNot:Attribute.HasDefaultValue==%><%==SIM:If:Attribute.OriginalType.HasClassifier==%><%==SIM:Attribute.OriginalType.Classifier.Name==%>()<%==SIM:ElseIf:Attribute.OriginalType.IsString==%>""<%==SIM:ElseIf:Attribute.OriginalType.IsNumber==%>0<%==SIM:ElseIf:Attribute.OriginalType.IsBoolean==%>False<%==SIM:Else==%>None<%==SIM:EndIf==%><%==SIM:EndIf==%><%==SIM:EndIf==%>    
			<%==SIM:Line==%>
	<%==SIM:EndIf==%><%==SIM:EndFor==%>
	<%==SIM:EndBlock==%>
	<%==SIM:Line==%>

	<%==SIM:ForEach:Element.AllAttributes==%>
        <%==SIM:If:Attribute.HasStereotype("property")==%>
            <%==SIM:ForEach:Attribute.DocumentationLines==%>
                #<%==SIM:DocumentationLine==%>
                <%==SIM:Line==%>
            <%==SIM:EndFor==%>

            @property<%==SIM:Line==%>
            def <%==SIM:Attribute.Name==%>(self):
            <%==SIM:StartBlock==%><%==SIM:Line==%>
                <%==SIM:If:IsEmpty(Attribute.GetBody)==%>
                    return self.<%==SIM:Attribute.Visibility==%><%==SIM:Attribute.Name==%>;<%==SIM:Line==%>
                <%==SIM:Else==%>
                    <%==SIM:Attribute.GetBody==%>                    
                <%==SIM:EndIf==%>
            <%==SIM:EndBlock==%>
            <%==SIM:Line==%>

            @<%==SIM:Attribute.Name==%>.setter<%==SIM:Line==%>
            <%==SIM:IfNot:Attribute.IsReadOnly==%>
                def <%==SIM:Attribute.Name==%>(self, value):
                <%==SIM:StartBlock==%><%==SIM:Line==%>
                    <%==SIM:If:IsEmpty(Attribute.SetBody)==%>
                        self.<%==SIM:Attribute.Visibility==%><%==SIM:Attribute.Name==%> = value;
                    <%==SIM:Else==%>
                        <%==SIM:Attribute.GetBody==%>                    
                    <%==SIM:EndIf==%>
                <%==SIM:EndBlock==%>
            <%==SIM:EndIf==%>
            
            
            <%==SIM:IfNot:IsLastItem==%><%==SIM:Line==%><%==SIM:EndIf==%>
            
        <%==SIM:EndIf==%>
    <%==SIM:EndFor==%>
    <%==SIM:Line==%>

	<%==SIM:ForEach:Element.Operations==%>    
	<%==SIM:ForEach:Operation.DocumentationLines==%>#<%==SIM:DocumentationLine==%><%==SIM:Line==%>
    <%==SIM:EndFor==%>
	def <%==SIM:Operation.Visibility==%><%==SIM:Operation.Name==%> (self<%==SIM:ForEach:Operation.Parameters==%>, <%==SIM:Parameter.Name==%><%==SIM:EndFor==%>):	
	<%==SIM:StartBlock==%><%==SIM:Line==%>
        <%==SIM:Line==%>		
		<%==SIM:EndBlock==%><%==SIM:Line==%>	


<%==SIM:EndFor==%>
<%==SIM:EndBlock==%>

File: software diagrams/SoftwareIdeasModeler-15-51/Data/Templates/Code/CSharp/Default/Interface.cs
<%==SIM:ExplicitWhitespacesOn==%>
<%==SIM:AutoIndentOn==%>

<%==SIM:Set:private=private==%><%==SIM:Set:protected=protected==%><%==SIM:Set:package=internal==%><%==SIM:Set:public=public==%>

<%==SIM:UserInput:ArrayType=Choice("Array type",Array,IEnumerable,ICollection,IList,List)==%>
<%==SIM:UserInput:Documentation=YesNo("Documentation")==%>

<%==SIM:Template:TypeTemplate(Typable)==%>
<%==SIM:If:Or(Typable.IsArray,Typable.ReturnIsArray)==%>
<%==SIM:If:Typable.IsUnique==%><%==SIM:If:Typable.IsOrdered==%>SortedSet<<%==SIM:Typable.Type==%>><%==SIM:Else==%>HashSet<<%==SIM:Typable.Type==%>><%==SIM:EndIf==%>
<%==SIM:Else==%>
<%==SIM:If:Typable.IsOrdered==%><%==SIM:If:Or(Eq(ArrayType,"Array"), IsEmpty(ArrayType))==%><%==SIM:Typable.Type==%>[]<%==SIM:Else==%><%==SIM:ArrayType==%><<%==SIM:Typable.Type==%>><%==SIM:EndIf==%>
<%==SIM:Else==%>ConcurrentBag<<%==SIM:Typable.Type==%>><%==SIM:EndIf==%>
<%==SIM:EndIf==%>
<%==SIM:Else==%>
<%==SIM:Typable.Type==%>
<%==SIM:EndIf==%><%==SIM:Space(1)==%>
<%==SIM:EndTemplate==%>

<%==SIM:Optional:System Usings==%>
    using System;<%==SIM:Line==%>
    using System.Collections.Generic;<%==SIM:Line==%>
    using System.Collections.Concurrent;<%==SIM:Line==%>
<%==SIM:EndOptional==%>
<%==SIM:ForEach:Imports==%>
    using <%==SIM:Import.Name==%>;
    <%==SIM:Line==%>
<%==SIM:EndFor==%>
<%==SIM:Line==%>

namespace <%==SIM:Element.Namespace==%><%==SIM:Line==%>
{<%==SIM:StartBlock==%><%==SIM:Line==%>

    <%==SIM:If:Documentation==%><%==SIM:If:Element.HasDocumentation==%>
        /// <summary><%==SIM:Line==%>
    <%==SIM:ForEach:Element.SimpleHtmlDocumentationLines==%>
        /// <%==SIM:DocumentationLine==%><%==SIM:Line==%>
    <%==SIM:EndFor==%>
        /// </summary><%==SIM:Line==%>
    <%==SIM:EndIf==%><%==SIM:EndIf==%>
    
    <%==SIM:IfNot:IsEmpty(Element.Visibility)==%><%==SIM:Element.Visibility==%> <%==SIM:EndIf==%>interface <%==SIM:Element.Name==%>

    <%==SIM:If:Element.HasTemplateParameters==%><<%==SIM:ForEach:Element.TemplateParameters==%>
            <%==SIM:Parameter.Name==%>
            <%==SIM:IfNot:IsLastItem==%>, <%==SIM:EndIf==%>
        <%==SIM:EndFor==%>><%==SIM:EndIf==%>
    <%==SIM:If:Element.HasTemplateParameterWithKind==%>
        <%==SIM:ForEach:Element.TemplateParameters==%>
            <%==SIM:If:Parameter.HasKind==%> where<%==SIM:Parameter.Name==%> : <%==SIM:Parameter.Kind==%> <%==SIM:EndIf==%>
        <%==SIM:EndFor==%>
    <%==SIM:EndIf==%>

    <%==SIM:If:Element.HasInterfaces==%> : <%==SIM:ForEach:Element.Interfaces==%><%==SIM:Interface.Name==%><%==SIM:IfNot:IsLastItem==%>, <%==SIM:EndIf==%><%==SIM:EndFor==%><%==SIM:EndIf==%>

    <%==SIM:Line==%>
    {<%==SIM:StartBlock==%><%==SIM:Line==%>

        <%==SIM:ForEach:Element.AllAttributes==%>            
            <%==SIM:If:Documentation==%><%==SIM:If:Attribute.HasDocumentation==%>
                /// <summary><%==SIM:Line==%>
            <%==SIM:ForEach:Attribute.SimpleHtmlDocumentationLines==%>
                /// <%==SIM:DocumentationLine==%><%==SIM:Line==%>
            <%==SIM:EndFor==%>
                /// </summary><%==SIM:Line==%>
            <%==SIM:EndIf==%><%==SIM:EndIf==%>
                
            <%==SIM:TypeTemplate(Attribute)==%><%==SIM:Space==%>
            <%==SIM:Attribute.Name==%>{ get; <%==SIM:IfNot:Attribute.IsReadOnly==%>set;<%==SIM:EndIf==%> }
            <%==SIM:IfNot:IsLastItem==%><%==SIM:Line==%><%==SIM:EndIf==%>
                
        <%==SIM:EndFor==%>

        <%==SIM:If:Element.HasAttributes==%><%==SIM:If:Element.HasOperations==%><%==SIM:Line(2)==%><%==SIM:EndIf==%><%==SIM:EndIf==%>

        <%==SIM:ForEach:Element.Operations==%>
            <%==SIM:If:Documentation==%><%==SIM:If:Operation.HasDocumentation==%>
                /// <summary><%==SIM:Line==%>
            <%==SIM:ForEach:Operation.SimpleHtmlDocumentationLines==%>
                /// <%==SIM:DocumentationLine==%><%==SIM:Line==%>
            <%==SIM:EndFor==%>
                /// </summary><%==SIM:Line==%>
            <%==SIM:ForEach:Operation.Parameters==%>
                /// <param name="<%==SIM:Parameter.Name==%>"><%==SIM:ToSingleLine(Parameter.Documentation)==%></param><%==SIM:Line==%>
            <%==SIM:EndFor==%><%==SIM:Line==%>
            <%==SIM:EndIf==%><%==SIM:EndIf==%>

            <%==SIM:If:Operation.HasReturnType==%><%==SIM:TypeTemplate(Operation)==%><%==SIM:EndIf==%><%==SIM:IfNot:Operation.HasReturnType==%>void<%==SIM:EndIf==%> <%==SIM:Operation.Name==%> (<%==SIM:ForEach:Operation.Parameters==%><%==SIM:Parameter.Type==%> <%==SIM:Parameter.Name==%><%==SIM:IfNot:IsLastItem==%>, <%==SIM:EndIf==%><%==SIM:EndFor==%>);            

            <%==SIM:Line==%>
        <%==SIM:EndFor==%>

    <%==SIM:EndBlock==%><%==SIM:Line==%>
    }

    <%==SIM:EndBlock==%><%==SIM:Line==%>
}